### PR TITLE
rgw: fix error handling in RGWBulkDelete::execute.

### PR DESCRIPTION
**Pull request by rzarzynski**
This is continuation of PR #8092.

### CODE REVIEW

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -4550,18 +4548,19 @@ void RGWBulkDelete::pre_exec()
 
 void RGWBulkDelete::execute()
 {
-  deleter = std::unique_ptr<Deleter>(new Deleter(store, s));
+  deleter = std::unique_ptr<Deleter>(new RGWBulkDelete::Deleter(store, s));
```
Comment (mattbenjamin) [inline src/rgw/rgw_op.cc]: it would be nice if, somehow, we did not need to make a new object (unique_ptr makes this pattern more visible, for me)

Comment (rzarzynski) [inline src/rgw/rgw_op.cc]: @mattbenjamin: I think we can easily avoid the dynamic allocation here with `boost::optional`. It's needed due to the lifetime of `deleter` crossing borders of `execute` and `send_response` methods. From the other side, we cannot make `deleter` in constructor of `RGWBulkDelete` as the `req_state` isn't initialized then. Anyway, `boost::optional` should work for us.

**Comment by rzarzynski** [comment]
@mattbenjamin: I've pushed the commit with dynalloc optimizations through `boost::optional`.

**Comment by mattbenjamin** [comment]
woot, I like this

File: src/rgw/rgw_op.cc (hunk 1 above)
Comment (cbodley) [inline src/rgw/rgw_op.cc]: another option is to use move semantics, i.e. the `RGWBulkDelete` constructor would default-construct a `Deleter`, and `execute()` would use move-assignment to replace it with a newly constructed `Deleter` once it has the `req_state`

**Comment by mattbenjamin** [comment]
@rzarzynski merged with auth-rework pr on wip-matt-testing

**Comment by mattbenjamin** [comment]
This had a segfault on my teuthology run, ping me.

**Comment by mattbenjamin** [comment]
@rzarzynski any update here?

**Comment by mattbenjamin** [comment]
@rzarzynski is this still active?  rebase?

**Comment by rzarzynski** [comment]
@mattbenjamin: yup, it's active.

The PR has been recently rebased and tested as a part of `wip-radek-testing`. Teuthology runs:
* http://pulpito.ceph.com/rzarzynski-2016-08-22_16:44:18-rgw-wip-radek-testing---basic-mira
* http://pulpito.ceph.com/rzarzynski-2016-08-29_11:18:44-rgw-wip-radek-testing---basic-mira

The failures seem to be unrelated to the branch's changes. I believe we can move forward.

**Comment by yehudasa** [comment]
@rzarzynski this needs a rebase, and also need to make sure there's a tracker issue opened for it and that it's referenced in the commit messages.

**Comment by rzarzynski** [comment]
@yehudasa: I will rebase and create the ticket soon. Also will try to replace the *in place factories* with `boost::optional::emplace()` that should be available thanks to the new Boost.

**Comment by rzarzynski** [comment]
@yehudasa: rebased.

**Comment by yehudasa** [comment]
@rzarzynski also need to make sure there's a tracker issue assigned to it and note it in the commit message. Was it tested after rebase?

**Comment by rzarzynski** [comment]
@yehudasa: I've just created the tracker issue and added the `Fixes` tag. I didn't retested the commit after rebasing as the merge conflict was trivial:
```
$ git diff
diff --cc src/rgw/rgw_op.cc
index 8ff83c6,b25afba..0000000
--- a/src/rgw/rgw_op.cc
+++ b/src/rgw/rgw_op.cc
@@@ -7,7 -6,8 +7,12 @@@
  
  #include <sstream>
  
++<<<<<<< HEAD
 +#include <boost/algorithm/string/predicate.hpp>
++=======
+ #include <boost/optional.hpp>
+ #include <boost/utility/in_place_factory.hpp>
++>>>>>>> rgw, optimization: avoid dynamic allocations of RGWBulkDelete::Deleter.
  
  #include "common/Clock.h"
  #include "common/armor.h"
```

**Comment by mattbenjamin** [comment]
@yehudasa, @liewegas this passed its teuthology run, w/by-hand 1-liner merge fix;  I plan to merge by hand

**Comment by mattbenjamin** [comment]
(this change has merged)

**Pull request closed without merging**